### DISCUSSION TITLE
SQL injection inside python f-strings

### DISCUSSION THREAD

**Discussion by ReubenM** [nvim-treesitter/nvim-treesitter#6073] [Q&A] [answered]
I'm wanting to use an SQL injection in python, but can't wrap my head around how to deal with f-string interpolation.

I've seen lots of examples that scan for a long list of SQL keywords, but that does not seem efficient, so I have the injection hook onto the `/*SQL*/` comment tag at the beginning of the string. This is minimal, and also makes it an optional feature.

Currently using:
```  
;extends
(string
  (string_content) @injection.content
    (#match? #injection.content "^\\/\\*SQL\\*\\/")
    (#set! injection.language "sql"))
```

Works great for normal strings, but breaks in f-strings at first interpolation variable. How do I set it so that the f-string interpolation variables act as wildcards for SQL parsing, in a way that it sees the entire query as a single sql query (not multiple queries separated by interpolation variables) without also joining together separate f-strings into a single query. 

So the following:

```python
foo = f'''/*SQL*/
  SELECT {x} FROM {y} WHERE {z} < 5
'''

bar = f'''/*SQL*/
  SELECT {x} FROM {y} WHERE {z} > 5
'''
```

I want to be able to have the syntax highlighting work across the complete f-string, so that treesitter sees this a 2 separate SQL queries, foo and bar.

Does treesitter even support handling this type of templating  interrupting an injection?

**Comment by bryankenote** [comment]
Is this similar to https://github.com/nvim-treesitter/nvim-treesitter/issues/3213?

**Reply by ReubenM** [reply]
Looks similar. If so, it might depend on neovim/neovim#17099 to be able to properly handle it.

**Reply by ReubenM** [reply]
I think this is fixed in neovim/neovim#24738

**Comment by ObserverOfTime** [comment]
You should **not** be interpolating SQL queries in the first place.

**Reply by ReubenM** [reply]
I'm well aware of sanitizing inputs by parameterizing queries . However, when building SQL queries, utilization of query parameters is limited to literals and identifiers. Anything else requires interpolation or concatenation of strings.

**Comment by gsiddhant159** [comment]
Any update on this? Were you able to find any solution/ workaround?

**Reply by ReubenM** [reply]
Unfortunately no. Right now I'm using:
```
;; extends
(
    ((string_content)(interpolation)*)+ @injection.content
    (#any-match? @injection.content "^\\/\\*SQL\\*\\/")
    (#set! injection.language "sql")
)
```

It kinda works, but the main problem is that the SQL string segments delimited by the interpolation chunks are each treated as separate SQL statements rather than being joined into a single SQL statement. So that throws a wrench in the SQL parser. There does exist a way of joining segments together, but it joins all segments across the entire document rather than just the local string.

**Comment by ReubenM** [comment, marked as answer]
For anybody who might come across this in searching, I figured I should update to make note that with recent treesitter changes over the past few months, the tree-sitter query injection I mentioned above is now working perfectly:

```
;; extends
(
    ((string_content)(interpolation)*)+ @injection.content
    (#any-match? @injection.content "^\\/\\*SQL\\*\\/")
    (#set! injection.language "sql")
)
```
